File: src/freight_fate/sim/trip_road_events.py
# ruff: noqa: F403,F405
"""Recurring road-event checks for trip simulation."""

from __future__ import annotations

from ..data.world import get_world
from .hos import is_night
from .trip_models import *
from .trip_route_helpers import *
from .weather import WeatherKind


class TripRoadEventMixin:
    def _traffic_pressure_intensity(self, mile: float, kind: str) -> float:
        leg_i, _ = self._leg_at_mile(mile)
        leg = self.route.legs[leg_i]
        intensity = 0.18
        if kind == "exit":
            intensity += 0.16
        elif kind == "route_merge":
            intensity += 0.20
        elif kind == "construction_merge":
            intensity += 0.34
        if self._near_city(mile):
            intensity += 0.22
        if leg.checkpoints:
            intensity += 0.12
        if self._rush_hour_traffic_bias(leg):
            intensity += 0.14
        if is_night(self.start_hour):
            intensity -= 0.06
        effects = self.weather.effects
        if effects.grip < 0.9:
            intensity += (0.9 - effects.grip) * 0.35
        if effects.visibility_mi < 3.0:
            intensity += (3.0 - effects.visibility_mi) * 0.04
        return max(0.0, min(0.95, intensity * self.hazard_scale))

    def _traffic_pressure_speed(self, mile: float, intensity: float) -> float:
        posted = self._corridor_limit_at(mile)
        return max(30.0, min(posted, posted - intensity * 26.0))

    def _place_traffic_pressures(self) -> list[TrafficPressure]:
        if self._is_facility_approach_route():
            # Merge/exit spacing pressure is highway language; city streets
            # get their pacing from per-street speed zones instead.
            return []
        pressures: list[TrafficPressure] = []

        def add(start: float, end: float, kind: str, direction: str, reason: str) -> None:
            start = max(0.0, start)
            end = min(self.total_miles, max(start + 0.2, end))
            intensity = self._traffic_pressure_intensity(start, kind)
            if intensity < TRAFFIC_PRESSURE_MIN_INTENSITY:
                return
            pressures.append(
                TrafficPressure(
                    start,
                    end,
                    kind,
                    direction,
                    intensity,
                    self._traffic_pressure_speed(start, intensity),
                    reason,
                )
            )

        for stop in self.stops:
            label = stop.exit_label or stop.spoken_name
            add(stop.at_mi - 2.0, stop.at_mi + 0.4, "exit", "right", f"exit traffic for {label}")
        for i, start in enumerate(self._leg_starts[1:], start=1):
            if self.route.legs[i - 1].highway != self.route.legs[i].highway:
                add(
                    start - 1.5,
                    start + 0.6,
                    "route_merge",
                    "right",
                    f"traffic merging for {self.route.legs[i].highway}",
                )
        for zone in self.zones:
            if zone.reason == "construction merge":
                add(
                    zone.start_mi,
                    zone.end_mi,
                    "construction_merge",
                    "left",
                    "construction taper traffic",
                )
        pressures.sort(key=lambda pressure: pressure.start_mi)
        return pressures

    def _check_tolls(self) -> None:
        for i, (start, leg) in enumerate(zip(self._leg_starts, self.route.legs, strict=True)):
            forward = self.route.cities[i] == leg.a
            for toll in leg.toll_events:
                offset = _stop_offset_for_direction(toll.at_mi, leg.miles, forward)
                at_mi = start + offset
                key = f"{i}:{toll.at_mi}:{toll.name}"
                if self.position_mi < at_mi or key in self._charged_tolls:
                    continue
                self._charged_tolls.add(key)
                if toll.amount <= 0:
                    self._emit(
                        TripEventKind.GPS_CUE,
                        f"{toll.method_label} entry recorded at {toll.name}; "
                        "toll will be billed at carrier settlement.",
                        toll=toll,
                    )
                    continue
                charge = TollCharge(toll, toll.amount)
                self.toll_charges.append(charge)
                estimate = "Estimated " if toll.estimated else ""
                self._emit(
                    TripEventKind.TOLL_CHARGED,
                    f"{toll.method_label} toll charged at {toll.name}: "
                    f"{estimate}{toll.amount:.0f} dollars, billed to carrier settlement.",
                    toll=toll,
                    amount=toll.amount,
                )

    @property
    def toll_expense(self) -> float:
        return sum(charge.amount for charge in self.toll_charges)

    def _check_cities(self) -> None:
        for i, start in enumerate(self._leg_starts):
            if i == 0 or i in self._announced_cities:
                continue
            if self.route.cities[i] == self.route.cities[i - 1]:
                # A same-city boundary is a surface-street segment change,
                # not a city passage; the turn cue already covers it.
                self._announced_cities.add(i)
                continue
            if self.position_mi >= start:
                self._announced_cities.add(i)
                prev = self.route.cities[i - 1]
                city = self.route.cities[i]
                nxt = self.route.cities[i + 1]
                leg = self.route.legs[i]
                world = get_world()
                city_state = world.cities[city].state
                prev_state = world.cities[prev].state
                crossing = f"Crossing into {city_state}. " if city_state != prev_state else ""
                self._emit(
                    TripEventKind.CITY_REACHED,
                    f"{crossing}Passing {city}, {city_state}. "
                    f"Continuing on {leg.highway} toward {nxt}.",
                )

    def _hazard_risk(self) -> float:
        """Chance of a hazard at each check; worse in fog and after dark."""
        vis = self.weather.effects.visibility_mi
        risk = 0.25 + (0.25 if vis < 2 else 0.0)
        if is_night(self.current_hour):
            risk += NIGHT_HAZARD_BONUS
        return risk * self.hazard_scale

    def _corridor_hazard_factor_at(self, mile: float) -> float:
        """Relative hazard-check frequency for the current corridor."""
        leg_i, _ = self._leg_at_mile(mile)
        leg = self.route.legs[leg_i]
        cls = _highway_class(leg.highway)
        factor = {"interstate": 1.05, "us_highway": 0.92}.get(cls, 0.82)
        factor += min(0.18, len(leg.checkpoints) * 0.06)
        region = self._region_at(mile)
        if region in _HOT_PATROL_REGIONS:
            factor += 0.12
        elif region in _COLD_PATROL_REGIONS:
            factor -= 0.12
        if self._near_city(mile):
            factor += 0.18
        return max(CORRIDOR_HAZARD_MIN_FACTOR, min(CORRIDOR_HAZARD_MAX_FACTOR, factor))

    def _next_hazard_check_interval_mi(self) -> float:
        base = self._rng.uniform(20, 60)
        return base / self._corridor_hazard_factor_at(self.position_mi)

    def _check_hazards(self, moved_mi: float) -> None:
        """Occasional road hazards that demand braking."""
        context = self.traffic_context()
        if (
            context is not None
            and context.closing_mph > 8.0
            and context.gap_seconds <= TRAFFIC_WARNING_GAP_S
            and self.position_mi >= self._traffic_warning_mi
        ):
            self._traffic_warning_mi = self.position_mi + 8.0
            # A lead vehicle blocks one lane: braking always works, and a
            # clear neighboring lane lets the player pass around it instead.
            # Reasons often end in "ahead" already, and a sub-tenth-mile gap
            # is spoken as "right ahead", not "0.0 miles ahead".
            reason = context.lead.reason.removesuffix(" ahead")
            where = (
                "right ahead" if context.gap_mi < 0.1 else f"{self._gap_text(context.gap_mi)} ahead"
            )
            self._emit(
                TripEventKind.HAZARD,
                f"Brake or change lanes! {reason.capitalize()} {where}.",
                deadline_s=2.5,
                traffic=context,
                dodgeable=True,
            )
            return
        self._hazard_check_mi -= moved_mi
        if self._hazard_check_mi > 0:
            return
        self._hazard_check_mi = self._next_hazard_check_interval_mi()
        if self._rng.random() < self._hazard_risk():
            choices = eligible_hazards(
                self.current_region,
                self.weather.current,
                self.terrain_at(self.position_mi),
                self.current_hour,
            )
            if not choices:
                return
            texts, weights = zip(*choices, strict=True)
            hazard = self._rng.choices(texts, weights)[0]
            dodgeable = hazard_is_dodgeable(hazard)
            call = "Brake or change lanes!" if dodgeable else "Brake now!"
            self._emit(
                TripEventKind.HAZARD,
                f"{call} {hazard[0].upper()}{hazard[1:]}.",
                deadline_s=(self._rng.uniform(3.0, 4.5) * self._visibility_reaction_factor()),
                dodgeable=dodgeable,
            )

    def _visibility_reaction_factor(self) -> float:
        """Low visibility shortens the normal hazard reaction slack."""
        vis = self.weather.effects.visibility_mi
        if vis >= 3.0:
            return 1.0
        return max(0.4, vis / 3.0)

    def _conditions_incident_text(self) -> str:
        """The traction-loss phrase for the current conditions."""
        kind = self.weather.current
        if kind == WeatherKind.SNOW:
            return "The trailer is sliding on the snow, too fast for the conditions."
        if kind in (WeatherKind.RAIN, WeatherKind.HEAVY_RAIN, WeatherKind.THUNDERSTORM):
            return "Hydroplaning on the wet road, too fast for the conditions."
        return "Losing traction, too fast for the conditions."

    def _check_conditions_speed(self, moved_mi: float) -> None:
        """Risk a traction-loss incident when driving too fast for slick roads."""
        eff = self.weather.effects
        over = self.truck.speed_mph - eff.safe_speed_mph
        if over <= CONDITIONS_SPEED_MARGIN_MPH or eff.grip >= CONDITIONS_GRIP_CEILING:
            self._conditions_check_mi = CONDITIONS_CHECK_MI
            return
        self._conditions_check_mi -= moved_mi
        if self._conditions_check_mi > 0:
            return
        self._conditions_check_mi = CONDITIONS_CHECK_MI
        severity = min(1.0, (over - CONDITIONS_SPEED_MARGIN_MPH) / 25.0)
        risk = severity * (1.0 - eff.grip) * CONDITIONS_INCIDENT_RISK * self.hazard_scale
        if self._cond_rng.random() < risk:
            self._emit(
                TripEventKind.HAZARD,
                f"Brake now! {self._conditions_incident_text()}",
                deadline_s=max(1.5, 2.5 * self._visibility_reaction_factor()),
            )

File: crates/ff-core/src/sim/trip_road_events.rs
//! Recurring road-event checks for trip simulation (port of
//! `freight_fate/sim/trip_road_events.py`, the former `TripRoadEventMixin`):
//! traffic pressures, tolls, city passages, hazards and too-fast-for-
//! conditions incidents.

use crate::data::world_models::py_capitalize;
use crate::pyfmt::{fmt_f, py_str_float};
use crate::sim::hos::is_night;
use crate::sim::trip::Trip;
use crate::sim::trip_models::*;
use crate::sim::trip_route_helpers::stop_offset_for_direction;
use crate::sim::weather::WeatherKind;
use crate::speech_text::{hazard_call, toll_charged, SpokenMessage, HAZARD_DODGE_CALL};

impl Trip {
    pub fn traffic_pressure_intensity(&self, mile: f64, kind: &str) -> f64 {
        let (leg_i, _) = self.leg_at_mile(mile);
        let leg = &self.route.legs[leg_i];
        let mut intensity = 0.18;
        match kind {
            "exit" => intensity += 0.16,
            "route_merge" => intensity += 0.20,
            "construction_merge" => intensity += 0.34,
            _ => {}
        }
        if self.near_city(mile) {
            intensity += 0.22;
        }
        if !leg.checkpoints().is_empty() {
            intensity += 0.12;
        }
        if self.rush_hour_traffic_bias(leg) != 0.0 {
            intensity += 0.14;
        }
        if is_night(self.local_start_hour()) {
            intensity -= 0.06;
        }
        let effects = self.weather.effects();
        if effects.grip < 0.9 {
            intensity += (0.9 - effects.grip) * 0.35;
        }
        if effects.visibility_mi < 3.0 {
            intensity += (3.0 - effects.visibility_mi) * 0.04;
        }
        (intensity * self.hazard_scale).clamp(0.0, 0.95)
    }

    pub fn traffic_pressure_speed(&self, mile: f64, intensity: f64) -> f64 {
        let posted = self.corridor_limit_at(mile);
        30.0_f64.max(posted.min(posted - intensity * 26.0))
    }

    pub fn place_traffic_pressures(&self) -> Vec<TrafficPressure> {
        if self.is_facility_approach_route() {
            // Merge/exit spacing pressure is highway language; city streets
            // get their pacing from per-street speed zones instead.
            return Vec::new();
        }
        let mut pressures: Vec<TrafficPressure> = Vec::new();
        let add = |pressures: &mut Vec<TrafficPressure>,
                   start: f64,
                   end: f64,
                   kind: &str,
                   direction: &str,
                   reason: String| {
            let start = start.max(0.0);
            let end = self.total_miles().min((start + 0.2).max(end));
            let intensity = self.traffic_pressure_intensity(start, kind);
            if intensity < TRAFFIC_PRESSURE_MIN_INTENSITY {
                return;
            }
            pressures.push(TrafficPressure {
                start_mi: start,
                end_mi: end,
                kind: kind.to_string(),
                direction: direction.to_string(),
                intensity,
                target_speed_mph: self.traffic_pressure_speed(start, intensity),
                reason,
            });
        };

        for stop in &self.stops {
            let label = if stop.exit_label.is_empty() {
                stop.spoken_name()
            } else {
                stop.exit_label.clone()
            };
            add(
                &mut pressures,
                stop.at_mi - 2.0,
                stop.at_mi + 0.4,
                "exit",
                "right",
                format!("exit traffic for {label}"),
            );
        }
        for i in 1..self.leg_starts.len() {
            let start = self.leg_starts[i];
            if self.route.legs[i - 1].highway != self.route.legs[i].highway {
                add(
                    &mut pressures,
                    start - 1.5,
                    start + 0.6,
                    "route_merge",
                    "right",
                    format!("traffic merging for {}", self.route.legs[i].highway),
                );
            }
        }
        for zone in &self.zones {
            if zone.reason == "construction merge" {
                add(
                    &mut pressures,
                    zone.start_mi,
                    zone.end_mi,
                    "construction_merge",
                    "left",
                    "construction taper traffic".to_string(),
                );
            }
        }
        pressures.sort_by(|a, b| {
            a.start_mi
                .partial_cmp(&b.start_mi)
                .expect("finite mileposts")
        });
        pressures
    }

    pub fn check_tolls(&mut self) {
        let legs = self.route.legs.clone();
        for (i, (start, leg)) in self.leg_starts.clone().iter().zip(legs.iter()).enumerate() {
            let forward = self.route.cities[i] == leg.a;
            for toll in leg.toll_events() {
                let offset = stop_offset_for_direction(toll.at_mi, leg.miles, forward);
                let at_mi = start + offset;
                let key = format!("{i}:{}:{}", py_str_float(toll.at_mi), toll.name);
                if self.position_mi < at_mi || self.charged_tolls.contains(&key) {
                    continue;
                }
                self.charged_tolls.insert(key);
                if toll.amount <= 0.0 {
                    self.emit(
                        TripEventKind::GpsCue,
                        SpokenMessage::new(format!(
                            "{} entry recorded at {}; toll will be billed at carrier settlement.",
                            toll.method_label(),
                            toll.name
                        )),
                        TripEventData {
                            toll: Some(toll.clone()),
                            ..Default::default()
                        },
                    );
                    continue;
                }
                self.toll_charges.push(TollCharge {
                    event: toll.clone(),
                    amount: toll.amount,
                });
                self.emit(
                    TripEventKind::TollCharged,
                    toll_charged(
                        &toll.method_label(),
                        &toll.name,
                        &fmt_f(toll.amount, 0),
                        toll.estimated,
                    ),
                    TripEventData {
                        toll: Some(toll.clone()),
                        amount: Some(toll.amount),
                        ..Default::default()
                    },
                );
            }
        }
    }

    pub fn toll_expense(&self) -> f64 {
        // Fold from 0.0 rather than `sum()`: Rust's `Sum for f64` starts at
        // -0.0, so an empty list yields negative zero and `fmt_grouped`
        // renders it "-0" -- a screen reader then says "minus zero
        // dollars". Python's `sum([])` is a plain 0.
        self.toll_charges
            .iter()
            .map(|charge| charge.amount)
            .fold(0.0, |total, amount| total + amount)
    }

    pub fn check_cities(&mut self) {
        let starts = self.leg_starts.clone();
        for (i, start) in starts.iter().enumerate() {
            if i == 0 || self.announced_cities.contains(&i) {
                continue;
            }
            if self.route.cities[i] == self.route.cities[i - 1] {
                // A same-city boundary is a surface-street segment change,
                // not a city passage; the turn cue already covers it.
                self.announced_cities.insert(i);
                continue;
            }
            if self.position_mi >= *start {
                self.announced_cities.insert(i);
                let prev = self.route.cities[i - 1].clone();
                let city = self.route.cities[i].clone();
                let nxt = self.route.cities[i + 1].clone();
                let highway = self.route.legs[i].highway.clone();
                let world = self.world;
                let city_state = world
                    .cities
                    .get(&city)
                    .unwrap_or_else(|| panic!("KeyError: {city:?}"))
                    .state
                    .clone();
                let prev_state = world
                    .cities
                    .get(&prev)
                    .unwrap_or_else(|| panic!("KeyError: {prev:?}"))
                    .state
                    .clone();
                // The mapped boundary is authoritative when the route has
                // one: it is announced at the surveyed mile, so repeating it
                // here would say the same state line twice. Ambient lines
                // ride a queue now, so the prefix goes back to being the
                // duplicate it always was.
                let crossing = if city_state != prev_state
                    && !self.leg_maps_crossing_into(i - 1, &city_state)
                {
                    format!("Crossing into {city_state}. ")
                } else {
                    String::new()
                };
                let message = format!(
                    "{crossing}Passing {}, {city_state}. Continuing on {highway} toward {}.",
                    world.spoken_city(&city, Some(false)),
                    world.spoken_city(&nxt, None)
                );
                self.emit(
                    TripEventKind::CityReached,
                    SpokenMessage::new(message),
                    TripEventData::default(),
                );
            }
        }
    }

    /// Whether this leg carries a surveyed crossing into `state`, read the
    /// same way `build_navigation_cues` reads it, direction and all.
    pub fn leg_maps_crossing_into(&self, leg_index: usize, state: &str) -> bool {
        let Some(leg) = self.route.legs.get(leg_index) else {
            return false;
        };
        let forward = self.route.cities[leg_index] == leg.a;
        leg.state_crossings().iter().any(|crossing| {
            let into_state = if forward {
                &crossing.state
            } else {
                &crossing.from_state
            };
            into_state == state
        })
    }

    /// Chance of a hazard at each check; worse in fog and after dark.
    pub fn hazard_risk(&self) -> f64 {
        let vis = self.weather.effects().visibility_mi;
        let mut risk = 0.25 + if vis < 2.0 { 0.25 } else { 0.0 };
        if is_night(self.local_hour()) {
            risk += NIGHT_HAZARD_BONUS;
        }
        risk * self.hazard_scale
    }

    /// Relative hazard-check frequency for the current corridor.
    pub fn corridor_hazard_factor_at(&self, mile: f64) -> f64 {
        let (leg_i, _) = self.leg_at_mile(mile);
        let leg = &self.route.legs[leg_i];
        let cls = highway_class(&leg.highway);
        let mut factor = match cls {
            "interstate" => 1.05,
            "us_highway" => 0.92,
            _ => 0.82,
        };
        factor += 0.18_f64.min(leg.checkpoints().len() as f64 * 0.06);
        let region = self.region_at(mile);
        if HOT_PATROL_REGIONS.contains(&region.as_str()) {
            factor += 0.12;
        } else if COLD_PATROL_REGIONS.contains(&region.as_str()) {
            factor -= 0.12;
        }
        if self.near_city(mile) {
            factor += 0.18;
        }
        factor.clamp(CORRIDOR_HAZARD_MIN_FACTOR, CORRIDOR_HAZARD_MAX_FACTOR)
    }

    pub fn next_hazard_check_interval_mi(&mut self) -> f64 {
        let base = self.rng.uniform(20.0, 60.0);
        base / self.corridor_hazard_factor_at(self.position_mi)
    }

    /// Occasional road hazards that demand braking.
    pub fn check_hazards(&mut self, moved_mi: f64) {
        if self.is_facility_approach_route() {
            // A deadhead crawl down a facility access road is minutes long
            // at yard speeds; a "brake now" ambush there is noise.
            return;
        }
        if let Some(context) = self.traffic_context() {
            if context.closing_mph > 8.0
                && context.gap_seconds() <= TRAFFIC_WARNING_GAP_S
                && self.position_mi >= self.traffic_warning_mi
            {
                self.traffic_warning_mi = self.position_mi + 8.0;
                // A lead vehicle blocks one lane: braking always works, and a
                // clear neighboring lane lets the player pass around it.
                let reason = context
                    .lead
                    .reason()
                    .strip_suffix(" ahead")
                    .unwrap_or(context.lead.reason())
                    .to_string();
                let where_ = if context.gap_mi < 0.1 {
                    "right ahead".to_string()
                } else {
                    format!("{} ahead", self.gap_text(context.gap_mi))
                };
                // "Or change lanes" is only true advice where there is
                // somewhere to send it (playtest report, US-285, 2026-08-12).
                let call = if self.has_open_adjacent_lane_at(None) {
                    HAZARD_DODGE_CALL
                } else {
                    "Brake!"
                };
                let message = hazard_call(call, &format!("{} {where_}.", py_capitalize(&reason)));
                self.emit(
                    TripEventKind::Hazard,
                    message,
                    TripEventData {
                        deadline_s: Some(2.5),
                        traffic: Some(context),
                        dodgeable: Some(true),
                        name: Some(format!("the {reason}")),
                        ..Default::default()
                    },
                );
                return;
            }
        }
        self.hazard_check_mi -= moved_mi;
        if self.hazard_check_mi > 0.0 {
            return;
        }
        self.hazard_check_mi = self.next_hazard_check_interval_mi();
        if self.rng.random() < self.hazard_risk() {
            let choices = eligible_hazards(
                self.current_region(),
                self.weather.current,
                &self.terrain_at(Some(self.position_mi)),
                self.local_hour(),
            );
            if choices.is_empty() {
                return;
            }
            let weights: Vec<f64> = choices.iter().map(|(_, w)| *w).collect();
            let idx = self.rng.choices_indices_weighted(&weights, 1)[0];
            let hazard = choices[idx].0;
            let dodgeable = hazard_is_dodgeable(hazard);
            let call = if dodgeable {
                if self.has_open_adjacent_lane_at(None) {
                    HAZARD_DODGE_CALL
                } else {
                    "Brake!"
                }
            } else {
                "Brake now!"
            };
            let mut chars = hazard.chars();
            let first: String = chars
                .next()
                .map(|c| c.to_uppercase().collect())
                .unwrap_or_default();
            let body = format!("{first}{}.", chars.as_str());
            let message = hazard_call(call, &body);
            let deadline_s = self.rng.uniform(3.0, 4.5) * self.visibility_reaction_factor();
            self.emit(
                TripEventKind::Hazard,
                message,
                TripEventData {
                    deadline_s: Some(deadline_s),
                    dodgeable: Some(dodgeable),
                    name: Some(hazard_name(hazard).to_string()),
                    ..Default::default()
                },
            );
        }
    }

    /// Low visibility shortens the normal hazard reaction slack.
    pub fn visibility_reaction_factor(&self) -> f64 {
        let vis = self.weather.effects().visibility_mi;
        if vis >= 3.0 {
            return 1.0;
        }
        0.4_f64.max(vis / 3.0)
    }

    /// The traction-loss phrase for the current conditions.
    pub fn conditions_incident_text(&self) -> &'static str {
        let kind = self.weather.current;
        if self.truck.hydroplaning() {
            return "Hydroplaning, the tires are riding the water film.";
        }
        match kind {
            WeatherKind::Ice => "The trailer is sliding on the ice.",
            WeatherKind::Snow => "The trailer is sliding on the snow, too fast for the conditions.",
            WeatherKind::Rain | WeatherKind::HeavyRain | WeatherKind::Thunderstorm => {
                "Hydroplaning on the wet road, too fast for the conditions."
            }
            _ => "Losing traction, too fast for the conditions.",
        }
    }

    /// The short noun phrase a resolution line names this incident by.
    pub fn conditions_incident_name(&self) -> &'static str {
        let kind = self.weather.current;
        if self.truck.hydroplaning() {
            return "the hydroplaning";
        }
        match kind {
            WeatherKind::Ice => "the ice",
            WeatherKind::Snow => "the snow",
            WeatherKind::Rain | WeatherKind::HeavyRain | WeatherKind::Thunderstorm => {
                "the hydroplaning"
            }
            _ => "the loss of traction",
        }
    }

    /// Risk a traction-loss incident when driving too fast for slick roads.
    /// The truck's own grip decides, and actually hydroplaning counts as deep
    /// overspeed no matter what the sign says.
    pub fn check_conditions_speed(&mut self, moved_mi: f64) {
        let eff = self.weather.effects();
        let grip = self.truck.effective_grip();
        let over = self.truck.speed_mph() - eff.safe_speed_mph;
        let planing = self.truck.hydroplaning();
        if (over <= CONDITIONS_SPEED_MARGIN_MPH && !planing) || grip >= CONDITIONS_GRIP_CEILING {
            self.conditions_check_mi = CONDITIONS_CHECK_MI;
            return;
        }
        self.conditions_check_mi -= moved_mi;
        if self.conditions_check_mi > 0.0 {
            return;
        }
        self.conditions_check_mi = CONDITIONS_CHECK_MI;
        let mut severity = ((over - CONDITIONS_SPEED_MARGIN_MPH).max(0.0) / 25.0).min(1.0);
        if planing {
            severity = severity.max(0.7);
        }
        let risk = severity * (1.0 - grip) * CONDITIONS_INCIDENT_RISK * self.hazard_scale;
        if self.cond_rng.random() < risk {
            let message = hazard_call("Brake now!", self.conditions_incident_text());
            let deadline_s = 1.5_f64.max(2.5 * self.visibility_reaction_factor());
            let name = self.conditions_incident_name().to_string();
            self.emit(
                TripEventKind::Hazard,
                message,
                TripEventData {
                    deadline_s: Some(deadline_s),
                    name: Some(name),
                    ..Default::default()
                },
            );
        }
    }
}
